File: src/206_Reverse_Linked_List.rs
// https://leetcode.cn/problems/reverse-linked-list/

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut pre = None;
        let mut head = head;
        while let Some(mut node) = head {
            head = node.next.take();
            node.next = pre;
            pre = Some(node);
        }
        pre
    }
}
// 作者：tab-liu
// 链接：https://leetcode.cn/problems/reverse-linked-list/solution/206-fan-zhuan-lian-biao-by-tab-liu-hp7g/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: src/206_Reverse_Linked_List.c
// https://leetcode.cn/problems/reverse-linked-list/

#include <stdio.h>

// Definition for singly-linked list.
struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *reverseList(struct ListNode *head)
{
    struct ListNode *pre = NULL;
    struct ListNode *left = head;
    struct ListNode *right = NULL;

    while (left)
    {
        right = left->next;
        left->next = pre;
        pre = left;
        left = right;
    }

    return pre;
}

struct ListNode *reverseList(struct ListNode *head)
{
    if (!head || !head->next)
    {
        return head;
    }

    struct ListNode *others = reverseList(head->next);
    head->next->next = head;
    head->next = NULL;

    return others;
}